sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala

```diff
@@ -108,6 +108,36 @@ trait StateStore {
   def hasCommitted: Boolean
 }
 
+/** A versioned key-value store which is same as [[StateStore]], but write-protected. */
+abstract class ReadOnlyStateStore extends StateStore {
+  override def put(key: UnsafeRow, value: UnsafeRow): Unit = throwNotAllowed()
+
+  override def remove(key: UnsafeRow): Unit = throwNotAllowed()
+
+  override def commit(): Long = throwNotAllowed()
+
+  override def metrics: StateStoreMetrics = throwNotAllowed()
```

Comment:
Given read-only stateful operator doesn't require this I've marked as 'not allowed', but semantically it also makes sense to allow this. It sounds correct to have metrics for read-only state store. It's just not used as of now.